### ISSUE TITLE
Displaying images uploaded to dropbox in Safari unreliable

### ISSUE THREAD

**Issue by endymion0000** [gravityview/GravityView#1691]
Safari doesn't appear to like displaying images reliably from dl.dropboxusercontent.com rather than www.dropbox.com with a ?raw=1 on then end when they are uploaded via form and moved to dropbox. You recently added the display method a few months back for dropbox hosted images.

With the dl.dropboxusercontent.com path, I can visit my grid of entry results or individual entry pages in Safari and it will seem fine and If I login the images will also appear fine initially. Then logout and view the same pages and log back in again to view the same pages and eventually all the images will break with console errors (see attached). It's weird, which is why I called it unreliable since you have to come back and revisit pages, reload pages, log in and out to see the problem, but for me it always happens enough that I immediately noticed the issue and only in Safari as far as I've seen.

Now, I did some testing for a solution, although this of course is not necessarily the proper edits, but here's what I did to change the path that works without the same problem.

First I edit:

/gravityview/includes/fields/class-gravityview-field-fileupload.php

line 327
and add 1 line:

		$image_atts['validate_src'] = false;

This lets me pass ?raw=1 on the end. Very likely a better way but I was just doing whatever to get a test to work sinceby default it appears to not allow ?raw=1 on then end without this.

Then edit:

/gravityview/includes/plugin-and-theme-hooks/class-gravityview-plugin-hooks-gravity-forms-dropbox.php

lines 47-51 replace with:
		
		$output = $string;
		//$output = str_replace('www.dropbox.com', 'dl.dropboxusercontent.com', $output );
		$output = str_replace( '?dl=0', '', $output );
		if (strpos($output, "www.dropbox.com")) {
			$output .= "?raw=1";
		}
		return $output;

You can see here I stop the conversion of the dropbox domain and then make sure I add the ?raw=1 when it's the full path in $output. I noticed sometimes $output was just the extension.

This works to where I can't reproduce the issue anymore in Safari. There must be something different in the dl.dropboxusercontent.com server response or headers that are causing Safari to wig out when it comes to caches or something. Again, it's weird.

So. my recommendation is stick with www.dropbox.com and add ?raw=1 to the end rather than changing the domain since it works better for me and other browsers treat it the same as before as far as I see.
![Screen Shot 2022-03-12 at 10 08 12 AM](https://user-images.githubusercontent.com/7957532/158024720-ea1012cc-553e-4eb1-8cae-67df5f6d4b21.png)
![Screen Shot 2022-03-12 at 10 08 26 AM](https://user-images.githubusercontent.com/7957532/158024728-46730dac-e161-4a53-a1c0-efe26ae46bc3.png)

**Comment by endymion0000** [comment]
I started inspecting the response headers and there are indeed differences that may be what trips up Safari and makes one more reliable than the other.

Using these 2 links:

A: https://www.dropbox.com/s/sg8xq6pubpfez3j/1750312.jpg?raw=1
B: https://dl.dropboxusercontent.com/s/sg8xq6pubpfez3j/1750312.jpg

My first tests show:

A: https://securityheaders.com/?q=https%3A%2F%2Fwww.dropbox.com%2Fs%2Fsg8xq6pubpfez3j%2F1750312.jpg%3Fraw%3D1&followRedirects=on

A: Has a referrer policy.

B: https://securityheaders.com/?q=https%3A%2F%2Fdl.dropboxusercontent.com%2Fs%2Fsg8xq6pubpfez3j%2F1750312.jpg&followRedirects=on

B: Does not have a referrer policy.

Let's look at the first non-redirected response.

A: https://securityheaders.com/?q=https%3A%2F%2Fwww.dropbox.com%2Fs%2Fsg8xq6pubpfez3j%2F1750312.jpg%3Fraw%3D1

A: Returns a 301 Redirect as expected since ultimately the url ends up at: https://uc09ed9cfbcf420c3d3511ddbe7f.dl.dropboxusercontent.com/cd/0/inline/BhWuOYSvgiaRsr-YUs-NL2whf8ZR8Ltt81_KBlVu_8-7MCdvdb1TprzWyog3SMI9yh5g2AjLw-DS3QzeakZLZoqGERKD3b2AsBziwmrAwlVVw7UUMtV_LxDmha9IjMsEYCNQNxs6tMWkdnMLxKnPn_qla99UT2drllCa1MNFTm0h3A/file#

B: https://securityheaders.com/?q=https%3A%2F%2Fdl.dropboxusercontent.com%2Fs%2Fsg8xq6pubpfez3j%2F1750312.jpg

B: Here it returns a 200 OK response, even though when I was testing with the images embedded in the site it still redirected to the other url. So for some reasons this returns 200 while maybe doing some redirect behind the scenes (this may be the root of the problem, since if safari is just caching the initial 200 response it may not be getting the image data back when it returns in some instances. The weird.part.

Here's another site:

A: https://iplocation.io/http-server-header-check/https://www.dropbox.com/s/sg8xq6pubpfez3j/1750312.jpg%3Fraw=1

A: Shows an initial 301 Redirect and then a 302 Redirect and then the final 200 OK.

B: https://iplocation.io/http-server-header-check/https://dl.dropboxusercontent.com/s/sg8xq6pubpfez3j/1750312.jpg

B: As before just the 200 OK.

I've even been able to trigger the issue in Safari directly by going straight to https://dl.dropboxusercontent.com/s/sg8xq6pubpfez3j/1750312.jpg

and then coming back after a bit (without reloading, just going to the link again in a new window and I get the attached screen.

![Screen Shot 2022-03-12 at 7 23 07 PM](https://user-images.githubusercontent.com/7957532/158039864-229c6918-d205-4157-8f92-b31d1fd66766.png)

So, ultimately there appears to be some Safari Bug not getting the response it needs  or it's caching of the response when using dl.dropboxusercontent.com which makes the www.dropbox.com…?raw=1 types with consistent redirects more reliable.

I can see not wanting to have redirects, but until they fix Safari, I'M not seeing another option. Is there a set of filters that would even allow changing it to work with the raw links from the functions file since normally it appears that ?raw=1 on the end is disallowed.

Maybe do some PHP browser detection to deliver the raw links to Safari and maybe Webkit in general browsers only and everyone else as you have it now.

At minimum I need something that will survive plugin updates if adjustments to the core files are not made via filters.

Thanks for any help.

**Comment by endymion0000** [comment]
OK, in my previous edits for

/gravityview/includes/plugin-and-theme-hooks/class-gravityview-plugin-hooks-gravity-forms-dropbox.php

lines 47-51 replace with is now:

		$output = $string;
		$output = str_replace( '?dl=0', '', $output );
		if (isset($_SERVER['HTTP_USER_AGENT'])) {
			if (strpos($_SERVER['HTTP_USER_AGENT'], "Safari") && !strpos($_SERVER['HTTP_USER_AGENT'], "Chrome")) {
				if (strpos($output, "www.dropbox.com")) {
					$output .= "?raw=1";
				}
			} else {
				$output = str_replace('www.dropbox.com', 'dl.dropboxusercontent.com', $output );
			}
		} else {
			$output = str_replace('www.dropbox.com', 'dl.dropboxusercontent.com', $output );
		}
		return $output;

This brings in the browser detection I mention to do it just for Safari if the User Agent is provided.

**Comment by endymion0000** [comment]
Been looking around online and getting not exactly clear info but I do see recommendations to just use the www.dropbox.com and ?raw=1 format links rather than dl.dropboxusercontent.com links.

There's some talk that the dl.dropboxusercontent.com when generated directly are considered temporary links and I suppose when used in this way regenerate the link to the actual image after it expires. So the link is consistent, but may be the issue with Safari caching since it's caching whatever was temporary on the dl.dropboxusercontent.com and not getting the up-to-date response. I guess dl.dropboxusercontent.com is just a proxy where the redirects are hidden from the end user, whereas the other link the redirects are visible and Safari knows better how to deal with that.

So, maybe best to just use the www.dropbox.com and ?raw=1 links across the board instead. Thoughts?

**Comment by zackkatz** [comment]
@endymion0000 Thank you so much for your in-depth analysis of the issue!

We'll incorporate the fix into the next release of the plugin. Until then, please [download GravityView using this link](https://gravityview.co/wp-content/uploads/p3ycbhfn174uvu0u2nzuevml3/gravityview-2.14.2.1-c7e4f93fb.zip) and let us know how it works for you!

**Comment by endymion0000** [comment]
The plugin version you linked is installed and working so far without issue.

Your welcome, I try to provide as much information that I can to 
determine what's happening or at least what I found along the way and this one was an odd unexpected one with the Safari cache. Memory cache seemed fine, it's when it was accessed from disk cache later where it became an issue. Since this is for a government related site, I needed to make sure everything was working as we near launch. Thanks for the quick response on this and my previous issue with empty search fields.

There's been a couple of other things but none that warranted an immediate issue. One is the SEO integration with Yoast for a Single Entry to populate The Title and Description tags only works when using the View's path. If one adds a view onto a page via shortcode and load an entry from there, the SEO tags aren't populates as set.

Another with the same functionality, I was trying to use the code in the manual, albeit listed as unsupported at the bottom (https://docs.gravityview.co/article/57-customizing-urls), where I could change the entry id such as /38/ to something using merge data like /this-place-38/ which did work fine, however similar to the other, the SEO title and description would fail but this time rather than working at all, the merge tag in the title would just not translate but would output "{Location Name:3}" instead in  the title. So, this may be part of it being unsupported. I was trying this to get more unique to the entry words on the url rather than just entry ids. And the url stuff worked fine. just the entry page not translating the merge data where it failed which kind of defeated part of the purpose, getting more SEO relevant info on the Title and url.
